Tutorial 7: Oracles doc (#601)
o1-labs/docs2: docs/zkapps/tutorials/07-oracle.mdx

## Prerequisites

- Make sure you have the zkApp CLI installed:

    ```sh
    $ npm install -g zkapp-cli
    ```

- Ensure your environment meets the [Prerequisites](/zkapps/tutorials#prerequisites) for zkApp Developer Tutorials.

This tutorial has been tested with:

- [zkApp CLI](https://www.npmjs.com/package/zkapp-cli) version 0.13.0
- [o1js](https://www.npmjs.com/package/o1js) version 0.13.0

## High-Level Overview

In this tutorial, you build an oracle that retrieves data from a REST API and write a smart contract that consumes information from this oracle.

1. Retrieve data from a REST API that provides mock credit score information for two users: one with a high credit score and one with a low credit score. 
1. The smart contract consumes this information and allows users to prove their credit score is above a certain threshold (for example, higher than 700).

Using the smart contract, an end user can generate an attestation that their credit score is above a certain value. To maintain their privacy, the user can prove this fact to a third party without sharing the exact credit score or other personal information.

> This tutorial uses a mock credit score API as the data source and provides a foundation to create an oracle for any type of data. Just swap out the oracle's HTTP request to use any REST API or GraphQL API source.  

Reviewer: ```suggestion
This tutorial uses a mock credit score API as the data source and provides a foundation to create an oracle for any type of data. Just alter to code to query data from whatever source you need, a REST API, for example. 
```
The existing oracle code doesn't actually make an HTTP request to another source; the values are simply hard-coded.

1. The smart contract consumes this information and allows users to prove their credit score is above a certain threshold (for example, higher than 700).

Using the smart contract, an end user can generate an attestation that their credit score is above a certain value. To maintain their privacy, the user can prove this fact to a third party without sharing the exact credit score or other personal information.

This tutorial uses a mock credit score API as the data source and provides a foundation to create an oracle for any type of data. Just swap out the oracle's HTTP request to use any REST API or GraphQL API source.  

## How Oracles Work

Oracles connect blockchain smart contracts with the outside world
to get data on chain. 

Mina smart contracts run off-chain and make it possible to prove that the expected computation was run on private data without revealing the data itself. When the smart contract consumes data from a third-party source, you want to verify that this data is authentic and was provided by the expected source.

> Mina zkOracles do not allow a zkApp to consume data trustlessly from any HTTPS data source. The oracle design described in this tutorial is typically operated by the zkApp developer. The oracle fetches and signs the desired data, then a zkApp can consume this data and verify the signature to ensure that the data was provided by the expected source.

Author: ```suggestion
The [Mina roadmap](https://minaprotocol.com/mina-roadmap) includes zkOracles to allow a zkApp to consume data trustlessly from any HTTPS data source. The oracle design described in this tutorial is typically operated by the zkApp developer. The oracle fetches and signs the desired data, and then a zkApp can consume this data and verify the signature to ensure that the data was provided by the expected source.
```

});
```

To run the tests:

1. Save the `CreditScoreOracle.test.ts` file. 
1. Run `npm run test`.

> You might have to try it twice in order for all the tests to pass because you are calling the live API directly. 

Reviewer: ```suggestion
```

This line isn't need any more :)
